ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,670 @@
+			warn!(target: "engine", "on_seal_block: unexpected extra extra_data: {:?}", header);
+		}
+		// vanity
+		{
+			let mut vanity = header.extra_data()[0..SIGNER_VANITY_LENGTH - 1].to_vec();
+			vanity.resize(SIGNER_VANITY_LENGTH, 0u8);
+			seal.extend_from_slice(&vanity[..]);
+		}
+
+		// If we are building an checkpoint block, add all signers now.
+		if is_checkpoint {
+			seal.reserve(state.signers.len() * 20);
+			state.signers.iter().foreach(|addr| {
+				seal.extend_from_slice(&addr[..]);
+			});
+		}
+
+		header.set_extra_data(seal.clone());
+
+		// append signature onto extra_data
+		let (sig, _msg) = self.sign_header(&header)?;
+		seal.extend_from_slice(&sig[..]);
+		header.set_extra_data(seal.clone());
+
+		header.compute_hash();
+
+		// locally sealed block don't go through valid_block_family(), so we have to record state here.
+		let mut new_state = state.clone();
+		new_state.apply(&header, header.number() % self.epoch_length == 0)?;
+		new_state.calc_next_timestamp(&header, self.period);
+		self.block_state_by_hash.write().insert(header.hash(), new_state);
+
+		trace!(target: "engine", "on_seal_block: finished, final header: {:?}", header);
+
+		Ok(Some(header))
+	}
+
+	/// Clique doesn't require external work to seal, so we always return true here.
+	fn seals_internally(&self) -> Option<bool> {
+		Some(true)
+	}
+
+	/// Returns if we are ready to seal, the real sealing (signing extra_data) is actually done in `on_seal_block()`.
+	fn generate_seal(&self, block: &ExecutedBlock, parent: &Header) -> Seal {
+		let null_seal = vec!(encode(&NULL_MIXHASH.to_vec()), encode(&NULL_NONCE.to_vec()));
+
+		trace!(target: "engine", "tried to generate seal");
+
+		if block.header.number() == 0 {
+			trace!(target: "engine", "attempted to seal genesis block");
+			return Seal::None;
+		}
+
+		// if sealing period is 0, and not an checkpoint block, refuse to seal
+		if self.period == 0 {
+			if block.transactions.is_empty() && block.header.number() % self.epoch_length != 0 {
+				return Seal::None;
+			}
+			return Seal::Regular(null_seal);
+		}
+
+		// Check we actually have authority to seal.
+		if let Some(author) = self.signer.read().as_ref().map(|x| x.address()) {
+			// ensure the voting state exists
+			match self.state(&parent) {
+				Err(e) => {
+					warn!(target: "engine", "generate_seal: can't get parent state(number: {}, hash: {}): {} ",
+						  parent.number(), parent.hash(), e);
+					return Seal::None;
+				}
+				Ok(state) => {
+					// Are we authorized to seal?
+					if !state.is_authoirzed(&author) {
+						trace!(target: "engine", "generate_seal: Not authorized to sign right now.");
+						// wait for one third of period to try again.
+						thread::sleep(Duration::from_secs(self.period / 3 + 1));
+						return Seal::None;
+					}
+
+					let inturn = state.inturn(block.header.number(), &author);
+
+					let now = SystemTime::now();
+
+					let limit = match inturn {
+						true => state.next_timestamp_inturn.unwrap_or(now),
```

Comment:
`state.next_timestamp_inturn` and `state.next_timestamp_noturn` seem related. Do you think there's a way these can be rolled into an `enum` so we can match over those to get a value for `limit` instead?

Reply:
No.. however geth don't store those values and just calculate them on the fly which I chose not to do because put it in block state is easier to debug..